src/test/java/org/sagebionetworks/bridge/services/TemplateServiceTest.java

```diff
@@ -0,0 +1,575 @@
+        verify(mockTemplateDao).updateTemplate(template);        
+    }
+    
+    @Test(expectedExceptions = EntityNotFoundException.class)
+    public void updateTemplateFailsIfDeleted() { 
+        Template existing = Template.create();
+        existing.setStudyId(TEST_STUDY_IDENTIFIER);
+        existing.setDeleted(true);
+        when(mockTemplateDao.getTemplate(TEST_STUDY, GUID1)).thenReturn(Optional.of(existing));
+        
+        Template template = Template.create();
+        template.setStudyId(TEST_STUDY_IDENTIFIER);
+        template.setGuid(GUID1);
+        template.setDeleted(true);
+        
+        service.updateTemplate(TEST_STUDY, template);
+    }
+    
+    @Test(expectedExceptions = InvalidEntityException.class)
+    public void updateTemplateInvalid() {
+        Template template = Template.create();
+        template.setGuid(GUID1);
+        when(mockTemplateDao.getTemplate(TEST_STUDY, GUID1)).thenReturn(Optional.of(template));
+        
+        service.updateTemplate(TEST_STUDY, template);
+    }
+    
+    @Test
+    public void deleteTemplate() {
+        Template existing = Template.create();
+        existing.setStudyId(TEST_STUDY_IDENTIFIER);
+        existing.setGuid(GUID1);
+        existing.setTemplateType(EMAIL_ACCOUNT_EXISTS);
+        when(mockTemplateDao.getTemplate(TEST_STUDY, GUID1)).thenReturn(Optional.of(existing));
+
+        service.deleteTemplate(TEST_STUDY, GUID1);
+        
+        verify(mockTemplateDao).updateTemplate(templateCaptor.capture());
+        
+        Template persisted = templateCaptor.getValue();
+        assertTrue(persisted.isDeleted());
+        assertEquals(persisted.getModifiedOn(), TIMESTAMP);
+    }
+    
+    @Test(expectedExceptions = EntityNotFoundException.class)
+    public void deleteTemplateLogicallyDeleted() { 
+        Template existing = Template.create();
+        existing.setStudyId(TEST_STUDY_IDENTIFIER);
+        existing.setGuid(GUID1);
+        existing.setDeleted(true);
+        when(mockTemplateDao.getTemplate(TEST_STUDY, GUID1)).thenReturn(Optional.of(existing));
+
+        service.deleteTemplate(TEST_STUDY, GUID1);
+    }
+    
+    @Test(expectedExceptions = EntityNotFoundException.class)
+    public void deleteTemplateMissing() { 
+        when(mockTemplateDao.getTemplate(TEST_STUDY, GUID1)).thenReturn(Optional.empty());
+
+        service.deleteTemplate(TEST_STUDY, GUID1);
+    }
+    
+    @Test
+    public void deleteTemplatePermanently() {
+        Template existing = Template.create();
+        existing.setStudyId(TEST_STUDY_IDENTIFIER);
+        existing.setGuid(GUID1);
+        existing.setTemplateType(EMAIL_ACCOUNT_EXISTS);
+        when(mockTemplateDao.getTemplate(TEST_STUDY, GUID1)).thenReturn(Optional.of(existing));
+
+        service.deleteTemplatePermanently(TEST_STUDY, GUID1);
+
+        verify(mockTemplateDao).deleteTemplatePermanently(TEST_STUDY, GUID1);
+    }
```

Comment:
Should we also verify that we delete the criteria?

Reply:
added